sys/drv/bcm2835_rootdev.c

```diff
@@ -1,137 +1,45 @@
 #define KL_LOG KL_DEV
+#include <sys/errno.h>
 #include <sys/klog.h>
 #include <sys/bus.h>
 #include <sys/devclass.h>
+#include <sys/fdt.h>
+#include <sys/libkern.h>
 #include <aarch64/armreg.h>
 #include <dev/bcm2835reg.h>
-#include <sys/kmem.h>
-#include <sys/pmap.h>
+#include <dev/simplebus.h>
 
 /*
  * located at BCM2836_ARM_LOCAL_BASE
  * 32 local interrupts -- one per CPU but now we only support 1 CPU
- *
- * located at BCM2835_ARMICU_BASE
- * accessed by BCM2835_PERIPHERALS_BASE NOT by BCM2835_PERIPHERALS_BASE_BUS
- * 32 GPU0 interrupts -- 0x204 offset from address; 32 in table
- * 32 GPU1 interrupts -- 0x208 offset from address; 64 in table
- * 32 base interrupts -- 0x200 offset from address; 96 in table
  */
 
-#define NIRQ (BCM2835_NIRQ + BCM2836_NIRQ)
-
 typedef struct rootdev {
   rman_t rm;
   rman_t irq_rm;
-  intr_event_t *intr_event[NIRQ];
+  resource_t mem_local; /* ARM local */
+  intr_event_t *intr_event[BCM2836_NIRQ];
 } rootdev_t;
 
-/* BCM2836_ARM_LOCAL_BASE mapped into VA (4KiB). */
-static vaddr_t rootdev_local_handle;
-/* BCM2835_PERIPHERALS_BUS_TO_PHYS(BCM2835_ARM_BASE) mapped into VA (4KiB). */
-static vaddr_t rootdev_arm_base;
-
-/* Use pre mapped memory. */
-static int rootdev_bs_map(bus_addr_t addr, bus_size_t size,
-                          bus_space_handle_t *handle_p) {
-  if (addr >= BCM2836_ARM_LOCAL_BASE &&
-      addr < BCM2836_ARM_LOCAL_BASE + BCM2836_ARM_LOCAL_SIZE) {
-    *handle_p = rootdev_local_handle;
-    return 0;
-  }
-  if (addr >= BCM2835_PERIPHERALS_BUS_TO_PHYS(BCM2835_ARM_BASE) &&
-      addr <
-        BCM2835_PERIPHERALS_BUS_TO_PHYS(BCM2835_ARM_BASE) + BCM2835_ARM_SIZE) {
-    *handle_p = rootdev_arm_base;
-    return 0;
-  }
-  return generic_bs_map(addr, size, handle_p);
-}
-
-/* clang-format off */
-static bus_space_t *rootdev_bus_space = &(bus_space_t){
-  .bs_map = rootdev_bs_map,
-  .bs_read_1 = generic_bs_read_1,
-  .bs_read_2 = generic_bs_read_2,
-  .bs_read_4 = generic_bs_read_4,
-  .bs_write_1 = generic_bs_write_1,
-  .bs_write_2 = generic_bs_write_2,
-  .bs_write_4 = generic_bs_write_4,
-  .bs_read_region_1 = generic_bs_read_region_1,
-  .bs_read_region_2 = generic_bs_read_region_2,
-  .bs_read_region_4 = generic_bs_read_region_4,
-  .bs_write_region_1 = generic_bs_write_region_1,
-  .bs_write_region_2 = generic_bs_write_region_2,
-  .bs_write_region_4 = generic_bs_write_region_4,
-};
-/* clang-format on */
-
-static void enable_local_irq(int irq) {
-  assert(irq < BCM2836_INT_NLOCAL);
-  uint32_t reg = bus_space_read_4(rootdev_bus_space, rootdev_local_handle,
-                                  BCM2836_LOCAL_TIMER_IRQ_CONTROLN(0));
-  bus_space_write_4(rootdev_bus_space, rootdev_local_handle,
-                    BCM2836_LOCAL_TIMER_IRQ_CONTROLN(0), reg | (1 << irq));
-}
+#define in4(addr) bus_read_4(&rd->mem_local, (addr))
+#define out4(addr, val) bus_write_4(&rd->mem_local, (addr), (val))
```

Comment:
Using local variables in macros is easy way to fuck-up something.

Reply:
I agree. It's error-prone and looks pretty weird. However, it's local to this file and it's common for our drivers to use this convention (which is granted by @cahirwpz). @cahirwpz do we want to drop this convention?